File: deck_game_installer/proton.py
import os
import re
import struct
from pathlib import Path
from typing import List, Tuple, Optional

class ProtonManager:
    def __init__(self):
        self.steam_path = Path.home() / ".local/share/Steam"
        self.common_path = self.steam_path / "steamapps/common"
        self.compat_tools_path = self.steam_path / "compatibilitytools.d"
        self.compat_data_path = self.steam_path / "steamapps/compatdata"

    def get_available_proton_versions(self) -> List[str]:
        """Returns a list of available Proton version internal names."""
        versions = []
        
        # Scan common/ for Steam-installed Proton
        if self.common_path.exists():
            for d in self.common_path.iterdir():
                if d.is_dir() and d.name.lower().startswith("proton"):
                    # Convert folder name to internal name: "Proton - Experimental" -> "proton-experimental"
                    internal_name = self._folder_to_internal_name(d.name)
                    versions.append(internal_name)
        
        # Scan compatibilitytools.d/ for custom Proton (e.g., Proton-GE)
        if self.compat_tools_path.exists():
            for d in self.compat_tools_path.iterdir():
                if d.is_dir():
                    # For custom tools, check for compatibilitytool.vdf to get internal name
                    vdf_path = d / "compatibilitytool.vdf"
                    if vdf_path.exists():
                        internal_name = self._get_internal_name_from_vdf(vdf_path)
                        if internal_name:
                            versions.append(internal_name)
                    else:
                        # Fallback to folder name conversion
                        versions.append(self._folder_to_internal_name(d.name))

        # Sort: Experimental first, then others descending
        def sort_key(v):
            if "experimental" in v.lower():
                return (0, "")
            # Extract version numbers if possible
            version_match = re.search(r"(\d+\.\d+)", v)
            if version_match:
                return (1, version_match.group(1))
            return (2, v)

        return sorted(list(set(versions)), key=sort_key)

    @staticmethod
    def _folder_to_internal_name(folder_name: str) -> str:
        """Converts a Proton folder name to its Steam internal name.
        
        Examples:
            'Proton - Experimental' -> 'proton-experimental'
            'Proton 10.0' -> 'proton_10'
            'Proton 9.0' -> 'proton_9'
        """
        name_lower = folder_name.lower()
        if "experimental" in name_lower:
            return "proton-experimental"
        # Extract version number for stable releases
        match = re.search(r"proton\s*(\d+)(?:\.\d+)?", name_lower)
        if match:
            return f"proton_{match.group(1)}"
        # Fallback: convert spaces/dashes to underscores
        return name_lower.replace(" - ", "-").replace(" ", "_")

    def _get_internal_name_from_vdf(self, vdf_path: Path) -> Optional[str]:
        """Extracts the internal tool name from a compatibilitytool.vdf file."""
        try:
            import vdf
            with open(vdf_path, "r") as f:
                data = vdf.load(f)
            compat_tools = data.get("compatibilitytools", {}).get("compat_tools", {})
            # Return the first key (the internal name)
            if compat_tools:
                return next(iter(compat_tools.keys()))
        except Exception:
            pass
        return None

    def scan_prefix_for_executables(self, app_id: int) -> List[Tuple[str, Path]]:
        # Steam uses unsigned 32-bit AppID strings for the prefix directory name
        u32_unsigned = struct.unpack('<I', struct.pack('<i', app_id))[0] if app_id < 0 else app_id
        prefix_path = self.compat_data_path / str(u32_unsigned) / "pfx/drive_c"
        if not prefix_path.exists():
            return []

        executables = []
        exclude_dirs = {"windows", "common files", "internet explorer", "steam", "users"}
        exclude_patterns = [
            r"unins.*\.exe", r"uninst.*\.exe", 
            r"vcredist.*\.exe", r"directx.*\.exe", r"dxsetup\.exe",
            r"setup\.exe", r"install\.exe", r"installer\.exe",
            r".*redist.*\.exe", r".*crash.*reporter.*\.exe"
        ]

        for root, dirs, files in os.walk(prefix_path):
            # Prune excluded directories
            rel_root = Path(root).relative_to(prefix_path)
            if any(part.lower() in exclude_dirs for part in rel_root.parts):
                continue

            for file in files:
                if file.lower().endswith(".exe"):
                    # Check against exclude patterns
                    if any(re.match(pattern, file, re.IGNORECASE) for pattern in exclude_patterns):
                        continue
                    
                    full_path = Path(root) / file
                    # Generate a display name like "game.exe (Program Files/GameName)"
                    try:
                        display_rel = full_path.relative_to(prefix_path)
                        display_name = f"{file} ({display_rel.parent})"
                    except ValueError:
                        display_name = file
                        
                    executables.append((display_name, full_path))

        # Sort by modification time (newest first)
        executables.sort(key=lambda x: x[1].stat().st_mtime, reverse=True)
        return executables

    @staticmethod
    def to_windows_path(pfx_path: Path, full_path: Path) -> str:
        """Converts a Linux path within a prefix to a Windows-style path (C:\\...)"""
        try:
            rel = full_path.relative_to(pfx_path)
            return "C:\\" + str(rel).replace("/", "\\")
        except ValueError:
            return str(full_path)

File: internal/proton/proton.go
package proton

import (
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strconv"
	"strings"

	"deck-game-installer/internal/vdf"
)

type Manager struct {
	steamPath       string
	commonPath      string
	compatToolsPath string
	compatDataPath  string
}

func NewManager() *Manager {
	home, _ := os.UserHomeDir()
	steamPath := filepath.Join(home, ".local", "share", "Steam")
	return &Manager{
		steamPath:       steamPath,
		commonPath:      filepath.Join(steamPath, "steamapps", "common"),
		compatToolsPath: filepath.Join(steamPath, "compatibilitytools.d"),
		compatDataPath:  filepath.Join(steamPath, "steamapps", "compatdata"),
	}
}

func (m *Manager) GetAvailableProtonVersions() []string {
	versions := []string{}
	if entries, err := os.ReadDir(m.commonPath); err == nil {
		for _, e := range entries {
			if e.IsDir() && strings.HasPrefix(strings.ToLower(e.Name()), "proton") {
				versions = append(versions, folderToInternalName(e.Name()))
			}
		}
	}

	if entries, err := os.ReadDir(m.compatToolsPath); err == nil {
		for _, e := range entries {
			if !e.IsDir() {
				continue
			}
			vdfPath := filepath.Join(m.compatToolsPath, e.Name(), "compatibilitytool.vdf")
			if data, err := os.ReadFile(vdfPath); err == nil {
				if name := internalNameFromVDF(string(data)); name != "" {
					versions = append(versions, name)
					continue
				}
			}
			versions = append(versions, folderToInternalName(e.Name()))
		}
	}

	versions = unique(versions)
	sort.Slice(versions, func(i, j int) bool {
		vi := versions[i]
		vj := versions[j]
		if strings.Contains(strings.ToLower(vi), "experimental") {
			return true
		}
		if strings.Contains(strings.ToLower(vj), "experimental") {
			return false
		}
		return vi > vj
	})

	return versions
}

func folderToInternalName(name string) string {
	lower := strings.ToLower(name)
	if strings.Contains(lower, "experimental") {
		return "proton-experimental"
	}
	if m := regexp.MustCompile(`proton\s*(\d+)`).FindStringSubmatch(lower); len(m) > 1 {
		return "proton_" + m[1]
	}
	return strings.ReplaceAll(strings.ReplaceAll(lower, " - ", "-"), " ", "_")
}

func internalNameFromVDF(data string) string {
	root, err := vdf.Parse(data)
	if err != nil {
		return ""
	}

	compat := nestedMap(root, "compatibilitytools", "compat_tools")
	if compat == nil {
		return ""
	}
	for k := range compat {
		return k
	}
	return ""
}

func nestedMap(root map[string]any, keys ...string) map[string]any {
	current := root
	for _, k := range keys {
		v, ok := current[k]
		if !ok {
			return nil
		}
		next, ok := v.(map[string]any)
		if !ok {
			return nil
		}
		current = next
	}
	return current
}

func unique(values []string) []string {
	seen := map[string]struct{}{}
	var out []string
	for _, v := range values {
		if _, ok := seen[v]; ok {
			continue
		}
		seen[v] = struct{}{}
		out = append(out, v)
	}
	return out
}

func (m *Manager) ScanPrefixForExecutables(appID int32) []string {
	unsigned := uint32(appID)
	prefix := filepath.Join(m.compatDataPath, strconv.FormatUint(uint64(unsigned), 10), "pfx", "drive_c")
	if _, err := os.Stat(prefix); err != nil {
		return nil
	}

	excludeDirs := map[string]struct{}{
		"windows": {}, "common files": {}, "internet explorer": {}, "steam": {}, "users": {},
	}
	patterns := []*regexp.Regexp{
		regexp.MustCompile(`unins.*\.exe`),
		regexp.MustCompile(`uninst.*\.exe`),
		regexp.MustCompile(`vcredist.*\.exe`),
		regexp.MustCompile(`directx.*\.exe`),
		regexp.MustCompile(`dxsetup\.exe`),
		regexp.MustCompile(`setup\.exe`),
		regexp.MustCompile(`install\.exe`),
		regexp.MustCompile(`installer\.exe`),
		regexp.MustCompile(`.*redist.*\.exe`),
		regexp.MustCompile(`.*crash.*reporter.*\.exe`),
	}

	var executables []string
	_ = filepath.WalkDir(prefix, func(path string, d os.DirEntry, err error) error {
		if err != nil {
			return nil
		}
		if d.IsDir() {
			rel, _ := filepath.Rel(prefix, path)
			parts := strings.Split(strings.ToLower(rel), string(filepath.Separator))
			for _, p := range parts {
				if _, ok := excludeDirs[p]; ok {
					return filepath.SkipDir
				}
			}
			return nil
		}
		name := strings.ToLower(d.Name())
		if !strings.HasSuffix(name, ".exe") {
			return nil
		}
		for _, re := range patterns {
			if re.MatchString(name) {
				return nil
			}
		}
		executables = append(executables, path)
		return nil
	})

	sort.Slice(executables, func(i, j int) bool {
		iInfo, _ := os.Stat(executables[i])
		jInfo, _ := os.Stat(executables[j])
		if iInfo == nil || jInfo == nil {
			return executables[i] > executables[j]
		}
		return iInfo.ModTime().After(jInfo.ModTime())
	})

	return executables
}
